plutus-core/plutus-core/src/PlutusCore/Core/Instance/Scoping.hs

```diff
@@ -0,0 +1,102 @@
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies          #-}
+
+module PlutusCore.Core.Instance.Scoping () where
+
+import           PlutusCore.Check.Scoping
+import           PlutusCore.Core.Type
+import           PlutusCore.Name
+import           PlutusCore.Quote
+
+-- In the three instances below the added variable is always the last field of a constructor.
+-- Just to be consistent.
+
+instance tyname ~ TyName => Reference TyName (Type tyname uni) where
+    referenceVia reg tyname ty = TyApp NotAName ty $ TyVar (reg tyname) tyname
+
+instance tyname ~ TyName => Reference TyName (Term tyname name uni fun) where
+    referenceVia reg tyname term = TyInst NotAName term $ TyVar (reg tyname) tyname
+
+instance name ~ Name => Reference Name (Term tyname name uni fun) where
+    referenceVia reg name term = Apply NotAName term $ Var (reg name) name
+
+-- Kinds have no names, hence the simple instance.
+instance Scoping Kind where
+    establishScoping kind = pure $ NotAName <$ kind
+    collectScopeInfo _ = mempty
+
+instance tyname ~ TyName => Scoping (Type tyname uni) where
+    establishScoping (TyLam _ nameDup kind ty) = do
```

Comment:
Yeah, I could, but I prefer not to do anything automatically if adding a new constructor or changing an existing one can give you a misbehaving code without anyone noticing. What if we add a new binder? It's like with `Eq` instances: I would rather have [this nonsense](https://github.com/input-output-hk/plutus/blob/52b23b80dc46048214ea169933e79678f806aa2b/plutus-core/plutus-core/src/PlutusCore/Core/Instance/Eq.hs#L121-L130) than risk forgetting to update the function resulting in a broken `Eq` instance without any warning.

`collectScopeInfo` could work though, but for that we'd need to generically navigate the type/term to determine which constructors have names and where their annotations are, plus that would just break if we end up having constructors with multiple names. I'd rather keep this wall of boilerplate than spend a lot of time trying to write some clever general machinery for something that is very straightforward to deal with otherwise.

So for example I think it was worth implementing the inference machinery for type schemes of builtins as it's non-trivial and rather annoying to write them out manually, plus the sets of builtins tend to change a lot. But updating those `Scoping` instances in case constructors change is a tiny task requiring zero mindfulness (largely due to the nice `Applicative`/`Monoid` API) (unless the new constructor is something creepy like a mutually recursive `let` or whatever, but that requires manual care regardless).

So I'd much prefer to keep the boilerplate.

Reply:
You could follow my suggestion and keep it explicit by just having no-op cases for `processTerm` for the interesting cases. That would be just as explicit, but would get rid of the recursion boilerplate and the annotation wrangling.